curriculum/challenges/english/06-information-security-and-quality-assurance/advanced-node-and-express/set-up-a-template-engine.english.md

```diff
@@ -6,21 +6,28 @@ forumTopicId: 301564
-A template engine enables you to use static template files (such as those written in <em>Pug</em>) in your app. At runtime, the template engine replaces variables in a template file with actual values which can be supplied by your server, and transforms the template into a static HTML file that is then sent to the client. This approach makes it easier to design an HTML page and allows for displaying of variables on the page without needing to make an API call from the client.
-To set up <em>Pug</em> for use in your project, you will need to add it as a dependency first in your package.json. <code>"pug": "^0.1.0"</code>
-Now to tell Node/Express to use the templating engine you will have to tell your express <b>app</b> to <b>set</b> 'pug' as the 'view-engine'. <code>app.set('view engine', 'pug')</code>
-Lastly, you should change your response to the request for the index route to <code>res.render</code> with the path to the view <em>views/pug/index.pug</em>.
-If all went as planned, you should refresh your apps home page and see a small message saying you're successfully rending the Pug from our Pug file! Submit your page when you think you've got it right.
+As a reminder, this project is built upon the following starter project on <a href='https://glitch.com/edit/#!/remix/clone-from-repo?REPO_URL=https://github.com/freeCodeCamp/boilerplate-advancednode/'>Glitch</a>, or clone from <a href='https://github.com/freeCodeCamp/boilerplate-advancednode/'>GitHub</a>.
+A template engine enables you to use static template files (such as those written in <em>Pug</em>) in your app. At runtime, the template engine replaces variables in a template file with actual values which can be supplied by your server. Then it transforms the template into a static HTML file that is sent to the client. This approach makes it easier to design an HTML page and allows for displaying variables on the page without needing to make an API call from the client.
+To set up <em>Pug</em> for use in your project, you will need to add it as a dependency in your package.json. Don't forget to add the name of the package and the version. Use the package.json for some examples of the correct syntax.
+
+Express needs to know which template engine you are using. We will use the <b>set</b> method to assign 'pug' as the <b>'view-engine'</b>. <code>app.set('view engine', 'pug')</code>
+
+Your page will not load until you correctly <em>render</em> the index file in our <em>'views/pug'</em> directory. Using the "Show" menu at the top of the Glitch page will allow you to view your app next to your code. This is helpful to see if your page is loading as expected. If you are seeing the page with a message "Pug template is not defined.", this indicates your routing is incomplete on the root path ('/'). You can get a <em>response</em> using a relative path to point the root ('/') to the index file.
```

Comment:
```suggestion
Your page will not load until you correctly <em>render</em> the index file in our <em>'views/pug'</em> directory. Using the "Show" menu at the top of the Glitch page will allow you to view your app next to your code. This is helpful to see if your page is loading as expected. If you are seeing the page with a message "Pug template is not defined.", you need to change your response to the request for the <code>/</code> route to render the file in <code>views/pug/index.pug</code>. <code>res.render()</code> takes a string of a file path an argument. The path can be a relative path (relative to views), or an absolute path, and does not require an extension. 
```
I liked @scissorsneedfoodtoo's wording, so I incorporated it here.

Reply:
I don't necessarily agree with the hand holding once we're into "advanced" coursework. This should be the beginning of pushing a camper to figuring out how to progress with reading docs and such.  What do you think?

Reply:
I think either we tell them what to do, or we point them to the docs so they can figure it out.  It's a little too much to expect people to realise that <em>render</em> refers to `res.render()`.

How about:

> Lastly, you should change your response to the request for the `/` route to use <code>res.render</code> to render the <code>index.pug</code> file from the <code>/views/pug/</code> directory.  The documentation for <code>res.render</code> can be found [here](https://expressjs.com/en/api.html#res.render) and the <code>view</code> path is relative to <code>/views</code>.
> 
>  Once you are successful your app will stop showing the message "Pug template is not defined." and will display "Looks like this page is being rendered from Pug into HTML!" instead.

It's a good point about "advanced" coursework, so if we want to introduce Glitch features (like 'Show') that should be done much earlier in the curriculum.

Reply:
By the time they reach the advanced node section, they should have a good grasp on reading docs and exploring a bit. They've already gone through express setup with the basic node curriculum. Hopefully by then they've explored some documentation too. I'm not in favor of spoon feeding this late in the course.

The glitch feature is already introduced in the first express lesson on the basic node coursework. [Working Express Server](https://www.freecodecamp.org/learn/apis-and-microservices/basic-node-and-express/start-a-working-express-server) 
>Note: Be sure that your code works by looking at the logs, then see the results in your browser by clicking the ‘Show Live’ button if you are using Glitch.

Reply:
As always, great discussion here, and @jeremyfiel brings up a good point about this being quite advanced in the curriculum. It's true that we don't want to spoon feed campers the correct answer and tell them exactly what to write and make it easy to copy and paste the answer in. But at the same time we want to keep them on the golden path as much as possible.

I'm in favor of @ojeytonwilliams's text above. It explains enough so campers won't need to go to the docs and look around but leaves that as in option for those who want to dive deeper into the `render` method.

Reply:
This text https://github.com/freeCodeCamp/freeCodeCamp/pull/38572#discussion_r412317698?  If so, I'll edit my review suggestion.

Reply:
@scissorsneedfoodtoo without giving them the actual method, do you think the tests are enough information not to spoon feed them this answer?  By this time, they've just completed the Chai testing curriculum prior to starting advanced node so they should be very familiar with gathering information from the test results. 
>text: Use the correct ExpressJS method to render the index page from the response.

If you are still leaning towards more descriptive text, please make a few grammatic changes: 
>you need to change your response to the request for the <code>/</code> route to render the file in <code>views/pug/~~index.pug~~</code>. <code>res.render()</code> takes a string of a file path **as an** argument. The path can be a relative path (relative to views), or an absolute path, and does not require **a file** extension. 

I bolded the changes.  I don't have access to modify the commit suggestion above.

Reply:
@jeremyfiel, I went through the Chai testing curriculum and think that those descriptions are a bit too vague at times. Ideally each description would explain more about each method that's introduced. That's something we're trying to improve on in the new testing projects for the project based curriculum.

Sorry @ojeytonwilliams, yes, I meant the text from https://github.com/freeCodeCamp/freeCodeCamp/pull/38572#discussion_r412317698. I think it strikes a good balance of being descriptive without giving the answer away. But I also like elements of the descriptive text in https://github.com/freeCodeCamp/freeCodeCamp/pull/38572#discussion_r412950966.

Perhaps we can combine the two and get the best of both worlds. Something like:

```
Lastly, you should change your response to the request for the `/` route to use <code>res.render</code> to render the <code>index.pug</code> file in the <code>/views/pug/</code> directory.  `res.render()` takes a string of a file path as an argument. The path can be a relative path (relative to views), or an absolute path, and does not require a file extension.

Once you are successful your app will stop showing the message "Pug template is not defined." and will display "Looks like this page is being rendered from Pug into HTML!" instead.
```

There's enough information there that we probably don't need the link to the docs.

Reply:
it's all ready for you @scissorsneedfoodtoo !  thanks again!